examples/smart_contract.ts

```diff
@@ -0,0 +1,277 @@
+import * as path from "http://localhost:5646/@local/deps/std/path.ts"
+import * as C from "http://localhost:5646/@local/mod.ts"
+import * as T from "http://localhost:5646/@local/test_util/mod.ts"
+import * as U from "http://localhost:5646/@local/util/mod.ts"
+import {
+  $contractsApiCallArgs,
+  $contractsApiCallResult,
+  $contractsApiInstantiateArgs,
+  $contractsApiInstantiateResult,
+} from "../frame_metadata/Contract.ts"
+
+const configFile = getFilePath("smart_contract/zombienet.toml")
+const zombienet = await T.zombienet.start(configFile)
+const client = zombienet.clients.byName["collator01"]!
+
+const contract = await getContract(
+  getFilePath("smart_contract/flipper.wasm"),
+  getFilePath("smart_contract/metadata.json"),
+)
+
+const contractAddress = U.throwIfError(await instantiateContractTx().run())
+console.log("Deployed Contract address", U.ss58.encode(42, contractAddress))
+
+function instantiateContractTx() {
+  const constructor = findContractConstructorByLabel("default")!
+  const salt = Uint8Array.from(Array.from([0, 0, 0, 0]), () => Math.floor(Math.random() * 16))
+  const value = preSubmitContractInstantiateDryRunGasEstimate(constructor, contract.wasm, salt)
+    .next(({ gasRequired, result: { accountId } }) => {
+      // the contract address derived from the code hash and the salt
+      console.log("Derived contract address", U.ss58.encode(42, accountId))
+      return {
+        type: "instantiateWithCode",
+        value: 0n,
+        gasLimit: {
+          refTime: gasRequired.refTime,
+          proofSize: gasRequired.proofSize,
+        },
+        storageDepositLimit: undefined,
+        code: contract.wasm,
+        data: U.hex.decode(constructor.selector),
+        salt,
+      }
+    })
+  const tx = C.extrinsic(client)({
+    sender: T.alice.address,
+    call: C.Z.rec({
+      type: "Contracts",
+      value,
+    }),
+  })
+    .signed(T.alice.sign)
+  const finalizedIn = tx.watch(({ end }) =>
+    (status) => {
+      if (typeof status !== "string" && (status.inBlock ?? status.finalized)) {
+        return end(status.inBlock ?? status.finalized)
+      } else if (C.rpc.known.TransactionStatus.isTerminal(status)) {
+        return end(new Error())
+      }
+      return
+    }
+  )
+  return C.events(tx, finalizedIn).next((events) => {
+    const extrinsicFailed = events.some((e) =>
+      e.event?.type === "System" && e.event?.value?.type === "ExtrinsicFailed"
+    )
+    if (extrinsicFailed) {
+      return new Error("extrinsic failed")
```

Comment:
How might we determine the cause?

Reply:
created an `ExtrinsicFailed` error and added the extrinsic failed event as the error `cause`